runtime/memprof.c

```diff
@@ -334,264 +346,322 @@ struct tracked {
-               && Wosize_val(res) == 1);
-    t = &trackst.entries[*t_idx];
-    t->user_data = Field(res, 0);
-    if (Is_block(t->user_data) && Is_young(t->user_data) &&
-        *t_idx < trackst.young)
-      trackst.young = *t_idx;
-  }
-
-  if (t->promoted && !t->cb_promote_called) {
-    t->cb_promote_called = 1;
-    user_data = t->user_data;
-    t->user_data = Val_unit;
-    res = run_callback_exn(t_idx, Promote(tracker), user_data);
-    if (*t_idx == Invalid_index)
-      return res;
-    CAMLassert(!Is_exception_result(res) && Is_block(res) && Tag_val(res) == 0
-               && Wosize_val(res) == 1);
-    t = &trackst.entries[*t_idx];
-    t->user_data = Field(res, 0);
-    if (Is_block(t->user_data) && Is_young(t->user_data) &&
-        *t_idx < trackst.young)
-      trackst.young = *t_idx;
-  }
-
-  if (t->deallocated && !t->cb_dealloc_called) {
-    value cb = (t->promoted || !t->alloc_young) ?
-      Dealloc_major(tracker) : Dealloc_minor(tracker);
-    t->cb_dealloc_called = 1;
-    user_data = t->user_data;
-    t->user_data = Val_unit;
-    res = run_callback_exn(t_idx, cb, user_data);
-    /* [t] is invalid, but we do no longer use it. */
-    CAMLassert(*t_idx == Invalid_index);
-    CAMLassert(Is_exception_result(res) || res == Val_unit);
-    return res;
-  }
+static value run_alloc_callback_exn(uintnat t_idx)
+{
+  struct tracked* t = &entries_local->t[t_idx];
+  value sample_info;
+
+  CAMLassert(Is_block(t->block) || Is_placeholder(t->block) || t->deallocated);
+  sample_info = caml_alloc_small(4, 0);
+  Field(sample_info, 0) = Val_long(t->n_samples);
+  Field(sample_info, 1) = Val_long(t->wosize);
+  Field(sample_info, 2) = Val_long(t->unmarshalled);
+  Field(sample_info, 3) = t->user_data;
+  return run_callback_exn(entries_local, t_idx,
+     t->alloc_young ? Alloc_minor(tracker) : Alloc_major(tracker), sample_info);
+}
 
-  return Val_unit;
+struct th_ctx_change_callback_status_data { intnat from, to; };
+static void th_ctx_change_cb_st(struct caml_memprof_th_ctx* ctx, void* data) {
+  struct th_ctx_change_callback_status_data *closure = data;
+  if(closure->from == ctx->callback_status) ctx->callback_status = closure->to;
 }
 
-/* Remove any deleted entries, updating callback and young */
-static void flush_deleted(void)
+/* Remove any deleted entries from [ea], updating [ea->young_idx] and
+   [callback_idx] if [ea == &entries_global]. */
+static void flush_deleted(struct entry_array* ea)
 {
-  uintnat i = trackst.delete, j = i;
-  while (i < trackst.len) {
-    if (!trackst.entries[i].deleted) {
-      if (trackst.entries[i].idx_ptr != NULL)
-        *trackst.entries[i].idx_ptr = j;
-      trackst.entries[j] = trackst.entries[i];
+  uintnat i, j;
+
+  if (ea == NULL) return;
+
+  j = i = ea->delete_idx;
+  while (i < ea->len) {
+    if (!ea->t[i].deleted) {
+      if (ea->t[i].callback_running) {
+        if (callback_status == i) callback_status = j;
```

Comment:
I think this is wrong if ea == entries_local. Can that happen?

Reply:
Well spotted, fixed.